bench/clblas_benchmark.cpp

```diff
@@ -0,0 +1,359 @@
+                                   context._queue(), 0, NULL, &event);
+        clWaitForEvents(1, &event);
+        clReleaseEvent(event);
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(axpy_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT alpha(2.4367453465);
+      MemBuffer<ScalarT> buf1(context, size);
+      MemBuffer<ScalarT> buf2(context, size);
+      cl_event event;
+      flops = benchmark<>::measure(no_reps, size * 1, [&]() {
+        clblasXaxpy<ScalarT>::func(size, alpha, buf1.dev(), 0, 1, buf2.dev(), 0,
+                                   1, 1, context._queue(), 0, NULL, &event);
+        clWaitForEvents(1, &event);
+        clReleaseEvent(event);
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(asum_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT vr;
+      MemBuffer<ScalarT, CL_MEM_WRITE_ONLY> buf1(context, size);
+      MemBuffer<ScalarT, CL_MEM_READ_ONLY> bufr(context, &vr, 1);
+      MemBuffer<ScalarT, CL_MEM_HOST_NO_ACCESS> scratch(context, size);
+      cl_event event;
+      flops = benchmark<>::measure(no_reps, size * 2, [&]() {
+        clblasXasum<ScalarT>::func(size, bufr.dev(), 0, buf1.dev(), 0, 1,
+                                   scratch.dev(), 1, context._queue(), 0, NULL,
+                                   &event);
+        clWaitForEvents(1, &event);
+        clReleaseEvent(event);
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(nrm2_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT vr;
+      MemBuffer<ScalarT, CL_MEM_WRITE_ONLY> buf1(context, size);
+      MemBuffer<ScalarT, CL_MEM_READ_ONLY> bufr(context, &vr, 1);
+      MemBuffer<ScalarT, CL_MEM_HOST_NO_ACCESS> scratch(context, 2 * size);
+      cl_event event;
+      flops = benchmark<>::measure(no_reps, size * 2, [&]() {
+        clblasXnrm2<ScalarT>::func(size, bufr.dev(), 0, buf1.dev(), 0, 1,
+                                   scratch.dev(), 1, context._queue(), 0, NULL,
+                                   &event);
+        clWaitForEvents(1, &event);
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(dot_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT vr;
+      MemBuffer<ScalarT, CL_MEM_WRITE_ONLY> buf1(context, size);
+      MemBuffer<ScalarT, CL_MEM_WRITE_ONLY> buf2(context, size);
+      MemBuffer<ScalarT, CL_MEM_READ_ONLY> bufr(context, &vr, 1);
+      MemBuffer<ScalarT, CL_MEM_HOST_NO_ACCESS> scratch(context, size);
+      cl_event event;
+      flops = benchmark<>::measure(no_reps, size * 2, [&]() {
+        clblasXdot<ScalarT>::func(size, bufr.dev(), 0, buf1.dev(), 0, 1,
+                                  buf2.dev(), 0, 1, scratch.dev(), 1,
+                                  context._queue(), 0, NULL, &event);
+        clWaitForEvents(1, &event);
+        clReleaseEvent(event);
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(iamax_bench) {
+    UNPACK_PARAM;
```

Comment:
I don't think this UNPACK_PARAM macro is needed, its not clear what it does and complicates debugging. Writing the original code is easier to follow